### ISSUE TITLE
Record Layer Protection - MAC

### ISSUE THREAD

**Issue by boaks** [tlswg/dtls-conn-id#25]
https://github.com/tlswg/dtls-conn-id/blob/master/draft-ietf-tls-dtls-connection-id.md#record-payload-protection

According RFC5246, the record layer protection is split into 3 sections
- NULL and stream (stream is not used for DTLS)
- CBC
- AEAD

in this draft, only the NULL is referenced
https://tools.ietf.org/html/rfc5246#section-6.2.3.1

If the CID value should be included in that MAC, I would expect, that this draft specifies how to do this for all three variants. By the way, I'm not sure, why the CID should be explicitly included. The CID "points" to the "keys" used to calculate the MAC, with a wrong CID, the wrong "key" will be used and that should cause a MAC violation anyway.

**Comment by ekr** [comment]
I can fix this text, but there are multiple CIDs pointing to the same keys, so we need to have the CID in the MAC

**Comment by boaks** [comment]
Sorry, I didn't get that. Why are "multiple CIDs" pointing to the same keys? 
Is "keys" referring to https://tools.ietf.org/html/rfc5246#section-6.1 

```
client write MAC key
server write MAC key
client write encryption key
server write encryption key
client write IV
server write IV
```

may be I miss something, if ip-address:port is used to identify the "keys", the MAC ensures, that the same keys are used. But the MAC itself is not based on the ip-address:port. or I'm wrong? If, I didn't miss something, then the cid is not required to be directly included in the MAC, as the ip-address:port. Instead is indirect validated with the MAC, as the ip-address:port would also be validated.

**Comment by ekr** [comment]
You can change CIDs during the connection. Thus, if it's not included in the MAC, you can substitute CIDs.

**Comment by boaks** [comment]
Sorry, I still can't see the point.
Do you mean, the CID could be exchanged during the transmission?
FMPOV this seems to be no issue, which could be fixed by including it explicit in the MAC.
The ip-address:port may also be exchanged, but are not included explicit in the MAC (this would fail on every NAT).  

Receiver side:
```
<CID1> => <mac keys1, enc keys1> 
<CID2> => <mac keys2, enc keys2> 
<CIDy> => no keys
``` 

Record `<CID1>` contents get protected by a MAC calculated with `<mac keys1>`.
If the `<CID1>` is exchanged during transmission with `<CIDy>`, this fails, because, no keys are mapped at the receiver.
If the `<CID1>` is exchanged during transmission with `<CID2>`, this fails, because, the receiver would try to validate the MAC using  `<mac keys2>`. 
For me, it's pretty the same, if you use the ip-address:port or CID to point to the keys, both are considered indirect and don't need to be included explicit. Or is there still a case, I missed?

**Comment by ekr** [comment]
You can have multiple valid CIDs for the same connection (and hence the same keys) at once. In this case, the attacker can swap them. Including the CID in the MAC protects against that.

I understand your analogy with IP address, but this is internal to TLS and might be used by the implementation, and therefore the safest practice is to protect it, like everything else in the header.

**Comment by boaks** [comment]
> You can have multiple valid CIDs

I can't see, that this is valid for this version of the draft!

> In DTLS, connection ids are exchanged at the beginning of the DTLS session only. There is no dedicated "connection id update" message that allows new connection ids to be established mid-session

If that is true, and the keys are generated in a way, which ensures that they are very likely to be unique for each session, how can I establish connections using different connections id but resulting in the same keys?

**Comment by boaks** [comment]
> You can have multiple valid CIDs for the same connection (and hence the same keys) at once.

May be an handshakes example, which results in that, will clarify it.

**Comment by ekr** [comment]
On Thu, Dec 6, 2018 at 6:31 AM Achim Kraus <notifications@github.com> wrote:

> You can have multiple valid CIDs
>
> I can't see, that this is valid for this version of the draft!
>
Yes, but it might be true with some extension in the future. Moreover,
authenticating the whole header is just good practice.

In DTLS, connection ids are exchanged at the beginning of the DTLS session
> only. There is no dedicated "connection id update" message that allows new
> connection ids to be established mid-session
>
> If that is true, and the keys are generated in a way, which ensures that
> they are very likely to be unique for each session, how can I establish
> connections using different connections id but resulting in the same keys?
>
Ah.

**Comment by boaks** [comment]
> Yes, but it might be true with some extension in the future. 

But isn't the future DTLS 1.3?

> Moreover, authenticating the whole header is just good practice.

But the CID is authenticated! Indirect! And, if it's not included in the MAC, then the changes in the implementations should be much smaller. Unfortunately, I had some other reported issues in californium, so I was just able to start to implement CID, but I couldn't finish it right now. I would really prefer, if this draft focus on fixing the NAT pain in DTLS 1.2 and leave the future to DTLS 1.3. But that's just my preference, and not really an argument.

**Comment by boaks** [comment]
> Ah.

Is for? 
1. my argument is that wrong, that no more comment is required
2. takes some time to think on it

**Comment by ekr** [comment]
On Thu, Dec 6, 2018 at 6:47 AM Achim Kraus <notifications@github.com> wrote:

> Ah.
>
> Is for?
>

typo.

**Comment by nmav** [comment]
Given that the CID replaces the IP:PORT tuple for TCP, not including in the MAC doesn't make the connection vulnerable to any attack that was not possible before. Indeed leaving it outside the calculations would make the DTLS1.2 implementation simpler and that's a plus.

**Comment by hannestschofenig** [comment]
Regarding the earlier point of not using the correct MAC definition I agree. Here is a proposed change: 
https://github.com/tlswg/dtls-conn-id/pull/29

My suggestion is to only focus on AEAD ciphers. 

Regarding the question whether the inclusion of the cid, cid_length, etc. is necessary or not I am not entirely sure myself either since I cannot see the attack. I do, however, understand that it is good practice to include more information in the MAC.

**Comment by boaks** [comment]
I still feel, that this extended/changed MAC will introduce some misunderstandings.

**Comment by hannestschofenig** [comment]
I tried to improve clarity in the description. If there is a way to improve the text please let me know. I fear that not making any changes to the additional data computation will require a formal analysis to ensure that it is safe to use.

**Comment by boaks** [comment]
> I tried to improve clarity in the description.

Yes, and I welcome  that very much.

FMPOV, it's open, what the "good practice" really includes. In TLS 1.2 it's the header and the "internal" seq_num to prevent repeats. With DTLS 1.2 it's parts of the header in order of the header, and, with the seq_num now also in the header, parts of the header **not** in order of that header. 
With this draft, not only the additional header field cid is added, it also adds the "internal" cid_length, and I'm wondering, why.

With https://github.com/tlswg/dtls-conn-id/pull/29#discussion_r246402825 I'm not sure, what the order of  the "good practice" should be.  

And agreeing with https://github.com/tlswg/dtls-conn-id/pull/29#discussion_r246403914 the other modes need to be defined also.

**Comment by boaks** [comment]
Will be fixed with #44
